### PR TITLE
Fix preloading of subset fonts in production environment

### PR DESCRIPTION
**Pull request by diox**
Fixes https://github.com/mozilla/addons-frontend/issues/10473

### CODE REVIEW

**Comment by diox** [comment]
I'll add a test for the plugin part.

File: src/amo/server/webpack-assets-fonts.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import fs from 'fs';
```
Comment (willdurand) [inline src/amo/server/webpack-assets-fonts.js:1]: the filename should be `WebpackAssetsFontsPlugin.js`

File: src/amo/server/webpack-assets-fonts.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+const pluginName = 'WebpackAssetsFontsPlugin';
```
Comment (willdurand) [inline src/amo/server/webpack-assets-fonts.js:3]: could be inlined

File: src/amo/server/webpack-assets-fonts.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          } = asset;
+
+          if (name.endsWith('.woff2')) {
+            subsetFonts[`./${sourceFilename}`] = `${publicPath}${name}`;
+          }
+        });
+
+        const data = JSON.parse(fs.readFileSync(webpackAssetsFileName));
+        data.assets = Object.assign(data.assets, subsetFonts);
+        fs.writeFileSync('webpack-assets.json', JSON.stringify(data));
```
Comment (willdurand) [inline src/amo/server/webpack-assets-fonts.js:30]:
```suggestion
        fs.writeFileSync(webpackAssetsFileName, JSON.stringify(data));
```

File: src/amo/server/webpack-assets-fonts.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            info: { sourceFilename },
+          } = asset;
+
+          if (name.endsWith('.woff2')) {
+            subsetFonts[`./${sourceFilename}`] = `${publicPath}${name}`;
+          }
+        });
+
+        const data = JSON.parse(fs.readFileSync(webpackAssetsFileName));
+        data.assets = Object.assign(data.assets, subsetFonts);
```
Comment (willdurand) [inline src/amo/server/webpack-assets-fonts.js:29]:
```suggestion
        data.assets = { ...data.assets, ...subsetFonts };
```

`Object.assign()` modifies the first argument so I think `data.assets` is updated _before_ being reassigned, which isn't what you want. In the past, we had to do `Object.assign({}, data.assets, subsetFonts)` but now we can use spread.

File: src/amo/server/webpack-assets-fonts.js (hunk 3 above)
Comment (willdurand) [reply src/amo/server/webpack-assets-fonts.js:30]: nit: I think we can use `fs-extra`: https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/writeJson-sync.md

**Review by willdurand — COMMENTED** [review]
I have minor comments only. The overall approach looks good.

File: src/amo/server/webpack-assets-fonts.js (hunk 4 above)
Comment (diox) [reply src/amo/server/webpack-assets-fonts.js:29]: thanks, I keep forgetting about spread...

File: src/amo/server/WebpackAssetsFontsPlugin.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+// This is a webpack plugin to add .woff2 fonts in webpack-assets.json that
+// webpack-isomorphic-tools generates. We can then use `assets` to reference
+// these fonts in server side rendered code to preload some font subsets, like
+// we can in development mode.
+
+export default class WebpackAssetsFontsPlugin {
+  constructor({ webpackAssetsFileName } = {}) {
+    if (!webpackAssetsFileName) {
+      throw new Error('The webpackAssetsFileName parameter cannot be empty');
```
Comment (willdurand) [inline src/amo/server/WebpackAssetsFontsPlugin.js:11]: I think it'd be slightly better to use the schema approach, which is the recommended way to validate webpack plugin options: https://webpack.js.org/contribute/writing-a-plugin/#basic-plugin-architecture

Comment (diox) [reply src/amo/server/WebpackAssetsFontsPlugin.js:11]: I got the idea from our own `sriDataPlugin.js` plugin :)

Comment (willdurand) [reply src/amo/server/WebpackAssetsFontsPlugin.js:11]: > I got the idea from our own `sriDataPlugin.js` plugin :)

yeah I was pretty sure. This plugin was written for webpack 1 I think.

**Review by willdurand — APPROVED** [review]
r+

**Pull request merged**